### ISSUE TITLE
Adding a privacy layer

### ISSUE THREAD

**Issue by saracubillas** [samtools/samtools#604]
SAM files can contains highly sensitive information that uniquely identifies an individual. 
The eXtensible Access Control Markup Language (XACML) [1] defined by OASIS defines a core XML schema for representing authorization and entitlement policies. 

I am developing a “privacy layer” that allows you to read and manipulate SAM files after evaluate access requests according to the rules defined. 
The idea is to use policies that specify rights and conditions associated with SAM files. 

> "Clearly, privacy concerns represent a formidable obstacle to assembling large human genomic databases and can delay (or derail) genome-wide association studies, which in turn could thwart advances in medicine and sub- sequent healthcare improvements." [2]

It is important research collaboration to specially and vigorously address the privacy issues that have been  identified.
<img width="818" alt="architectureschema" src="https://cloud.githubusercontent.com/assets/1815335/17737303/d1efef02-648c-11e6-8332-fbd20d5b7e93.png">
[1] XACML v3.0, https://www.oasis‐ open.org/committees/tc_home.php wg_abbrev=xacml#CURRENT  

[2]Whole Genome Sequencing: Revolutionary Medicine or Privacy Nightmare?. Ayday, Erman; De Cristofaro, Emiliano; Hubaux, Jean-Pierre; Tsudik. * Published in: IEEE Computer, vol. 48, num. 2, p. 58-66. * Los Alamitos: Ieee Computer Soc, 2015

**Comment by mp15** [comment]
Edited to fix the URL

**Comment by SamStudio8** [comment]
This issue seems to be outside the scope of the `samtools` project? From the diagram I understand that the proposed workflow essentially encapsulates the execution of `samtools` (but presumably this could really be any software, such as the GATK). Is there an action\request on the part of the `samtools` team here?

**Comment by saracubillas** [comment]
Yes, it encapsulate the execution of samtools, but it is a proof of concept, to see how privacy could be integrated with the SAM format. As you can see on the diagram there is a connection between policy and the SAM files, how it should be related? I still don’t now but maybe a solution could be attempt to introduce XACML on the optional tags of the  alignment section. Do you think that this work should be done outside the samtools project?

**Comment by SamStudio8** [comment]
Unless you need read-level access control, I imagine you would have optional tags included as part of the SAM/BAM header (with some sort of checksum?). Your question might be better suited to be opened as an issue on the https://github.com/samtools/hts-specs repository?

**Comment by jkbonfield** [comment]
There is room for discussion as to how the decryption process should be coupled with Samtools.

For example, one standard (and poor!) way that people handle sensitive data currently is to encrypt the entire file.  When querying it the entire file gets decrypted and cached locally prior to running samtools.   This then forbids any random access on an encrypted stream and it requires a local unencrypted file, which inherently becomes a security risk.  We need to consider what happens with temporary files; to ensure we use umask correctly, etc.

An alternative, and something very much in the hts-specs realm, is whether each bgzf block or each cram container could be encrypted independently, presumably with some salt to stop some of the attack forms.  This could make it possible to do random access on the encrypted files, perhaps even permitting remote browsing (although that's a whole new can of worms with the authorization).

Some data still leaks - the index for example either needs to be unencrypted (so coverage data can be gleaned) or for it to be encrypted but entirely decrypted prior to processing.  There are almost certainly timing attacks too, but I assume you don't want to try and protect it to that extreme.

**Comment by jkbonfield** [comment]
Infact by "salt" in the above statement, I realise it's trivial.  File offset + key => new encryption key for this bgzf block.  The bai index gives the block offset, the server has the key.  Combine both together, do the low level read, decrypt with offset+key, process as normal.

It doesn't sound that complex really; for the samtools/htslib side that is. :-)
